### DISCUSSION TITLE
Improve window resizing/scaling behavior

### DISCUSSION THREAD

**Discussion by SuperFromND** [ikemen-engine/Ikemen-GO#1876] [Engine development]
I recently [added the ability for Ikemen GO's window to be resized independently of game resolution](https://github.com/ikemen-engine/Ikemen-GO/commit/b3447d8d12edea66344f1f865156d6629e15fc17) similar to how MUGEN 1.1/SDL handles its game window. The problem is, the way I implemented it is quite literally just calling `gl.Viewport` to resize and offset the viewport to fit the window. This _works_, but it's not ideal and doesn't behave like how other programs would do it (taking the frame and projecting it onto a texture, then rendering that texture to the screen; doing it this way gives essentially free filtering as well as properly working with shaders, something the viewport approach breaks :<).

If I was just a bit more savvy with OpenGL, this could probably be pretty easy to do, but I figured I'd throw it out here in the discussions channel just in case someone in internet-land who knows OpenGL better stumbles across it. 👍

**Comment by SuperFromND** [comment]
Trying to improve it myself, this modded version of `EndFrame()` *almost* works:

```go
func (r *Renderer) EndFrame() {
	if sys.multisampleAntialiasing {
		gl.BindFramebuffer(gl.DRAW_FRAMEBUFFER, r.fbo_f)
		gl.BindFramebuffer(gl.READ_FRAMEBUFFER, r.fbo)
		gl.BlitFramebuffer(0, 0, sys.scrrect[2], sys.scrrect[3], 0, 0, sys.scrrect[2], sys.scrrect[3], gl.COLOR_BUFFER_BIT, gl.LINEAR)
	}

	x, y, resizedWidth, resizedHeight := sys.window.GetScaledViewportSize()

	postShader := r.postShaderSelect[sys.postProcessingShader]

	gl.Clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT)
	gl.UseProgram(postShader.program)
	gl.Disable(gl.BLEND)

	gl.ActiveTexture(gl.TEXTURE0)
	if sys.multisampleAntialiasing {
		gl.BindTexture(gl.TEXTURE_2D, r.fbo_f_texture.handle)
	} else {
		gl.BindTexture(gl.TEXTURE_2D, r.fbo_texture)
	}
	gl.Uniform1i(postShader.u["Texture"], 0)
	gl.Uniform2f(postShader.u["TextureSize"], float32(sys.scrrect[2]), float32(sys.scrrect[3]))

	gl.BindBuffer(gl.ARRAY_BUFFER, r.postVertBuffer)

	loc := r.modelShader.a["VertCoord"]
	gl.EnableVertexAttribArray(uint32(loc))
	gl.VertexAttribPointerWithOffset(uint32(loc), 2, gl.FLOAT, false, 0, 0)

	gl.DrawArrays(gl.TRIANGLE_STRIP, 0, 4)

	gl.DisableVertexAttribArray(uint32(loc))
	gl.Viewport(x, y, resizedWidth, resizedHeight)
	gl.BindFramebuffer(gl.READ_FRAMEBUFFER, r.fbo_texture)
	gl.BindFramebuffer(gl.DRAW_FRAMEBUFFER, 0)
	gl.BlitFramebuffer(0, 0, sys.scrrect[2], sys.scrrect[3], x, y, x+resizedWidth, y+resizedHeight, gl.COLOR_BUFFER_BIT, gl.LINEAR)
}
```

The result *almost* works, but exhibits pretty severe corruption (likely as a result of reading `r.fbo_texture` directly) and it doesn't scale at all if MSAA is turned on (giving solid black when attempting to scale, but working when only one dimension is resized).
![image](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/42b4aefe-5bde-4c97-9934-11cf988f4dd1)

**Reply by SuperFromND** [reply]
Just now found out removing the `gl.Clear` command at least somewhat mitigates the corruption, but it still remains. :<

According to RenderDoc, the actual underlying texture is fine, just the part where it gets drawn to the screen is malfunctioning. Which means either `gl.Viewport` or `gl.BlitFramebuffer` is the cause of the problem.

**Comment by SuperFromND** [comment]
#1895 should now address these problems. 👍

**Reply by Kiwi1444** [reply]
Is there any chance that IkemenGo will remember the position and size of the window?
